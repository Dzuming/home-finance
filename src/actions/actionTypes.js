const asyncActionType = type => ({
  PENDING: `${type}_PENDING`,
  SUCCESS: `${type}_SUCCESS`,
  ERROR: `${type}_ERROR`
});

const crudActionType = type => ({
  CREATE: `CREATE_${type}`,
  READ: `READ_${type}`,
  UPDATE: `UPDATE_${type}`,
  DELETE: `DELETE_${type}`
});

const draggedElements = type => ({
  READ: `READ_${type}`,
  SELECT: `SELECT${type}`,
  RESET: `RESET_${type}`,
  REMOVE: `REMOVE_${type}`
});

export const OPEN_SIDENAV = 'OPEN_SIDENAV';
export const AUTH_USER = 'AUTH_USER';
export const AUTH_REQUEST = 'AUTH_REQUEST';
export const REQUEST_LOGOUT = 'REQUEST_LOGOUT';
export const SUCCESS_LOGOUT = 'SUCCESS_LOGOUT';
export const SET_USER = 'SET_USER';
export const SET_DATE = 'SET_DATE';
export const API_REQUEST_GET = 'API_REQUEST_GET';
export const SET_CATEGORIES = 'SET_CATEGORIES';
export const API_REQUEST_POST = 'API_REQUEST_POST';
export const API_REQUEST_DELETE = 'API_REQUEST_DELETE';
export const API_REQUEST_PUT = 'API_REQUEST_PUT';
export const API_REQUEST = asyncActionType('API_REQUEST');
export const SPENDING_CRUD = crudActionType('SPENDING');
export const PROFIT_CRUD = crudActionType('PROFIT');
export const BUDGET_CRUD = crudActionType('BUDGET');
export const REVENUE_CRUD = crudActionType('REVENUE');
export const ASSUMPTION_CRUD = crudActionType('ASSUMPTION');
export const OVERALL_ASSUMPTIONS_CRUD = crudActionType('OVERALL_ASSUMPTIONS');
export const ASSUMPTION_TYPES = draggedElements('ASSUMPTION_TYPES');
export const CATEGORIES_DRAG = draggedElements('CATEGORIES');
