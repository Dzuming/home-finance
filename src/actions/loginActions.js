import * as types from './actionTypes';
import { getUser } from './userActions';
import env from '../../environments/config';

function requestLogin () {
  return {type: types.REQUEST_LOGIN};
}

function successLogin (accessToken) {
  setStorageToken(accessToken);
  return {type: types.SUCCESS_LOGIN};
}

function loginError (message) {
  return {type: types.LOGIN_FAILURE, message};
}

function authUser () {
  return {type: types.AUTH_USER};
}

function authenticateUser (credentials) {
  const user = credentials;
  let formData = new FormData();
  formData.append('grant_type', 'password');
  formData.append('client_id', '2');
  formData.append('client_secret', 'tbvtGeM45Jnk8IFRTVXFiBJrySgtU7pgDehuZoKG');
  formData.append('username', user.email);
  formData.append('password', user.password);
  formData.append('scope', '*');
  return dispatch => {
    dispatch(requestLogin());
    return fetch(`${env.api_url}/oauth/token`, {
      method: 'POST',
      body: formData
    }).then(response => {
      if (!response.ok) {
        throw Error(response.statusText);
      }
      return response.json();
    }).then((data) => {
      const accessToken = data.access_token;
      dispatch(successLogin(accessToken));
      dispatch(authUser());
      dispatch(getUser(user.email, accessToken));
    }).catch((error => dispatch(loginError(error))));
  };
}

function setStorageToken (token) {
  localStorage.setItem('token', token);
}

export function login (data) {
  // Note that the function also receives getState() which lets you choose what to
  // dispatch next. This is useful for avoiding a network request if a cached
  // value is already available.
  return (dispatch) => {
    // Dispatch a thunk from thunk!
    return dispatch(authenticateUser(data));
  };
}
