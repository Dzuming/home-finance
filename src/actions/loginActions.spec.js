import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import * as actions from './loginActions';
import * as types from './actionTypes';
import fetchMock from 'fetch-mock';
import expect from 'expect';
import env from '../../environments/config';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('async actions', () => {
  afterEach(() => {
    fetchMock.reset();
    fetchMock.restore();
  });

  it('creates SUCCESS_LOGIN when authentication has been done', () => {
    fetchMock
      .postOnce(`${env.api_url}/oauth/token`, {
        body: {token: '12345', user: 'John Doe'},
        headers: {'content-type': 'application/json'}
      });
    const expectedActions = [
      {type: types.REQUEST_LOGIN},
      {type: types.SUCCESS_LOGIN},
      {type: types.AUTH_USER},
      {type: types.SET_USER, user: 'John Doe'},
    ];
    const store = mockStore({token: '', user: []});
    const credentials = {
      email: 'test@test.com',
      password: 'secret'
    };
    return store.dispatch(actions.login(credentials)).then(() => {
      // return of async actions
      expect(store.getActions()).toEqual(expectedActions);
    });
  });
});
