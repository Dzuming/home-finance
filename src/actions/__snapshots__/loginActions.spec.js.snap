// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login actions should create authenticateUser action with snapshot 1`] = `
Object {
  "type": "AUTH_USER",
}
`;
